src/v/raft/replicate_batcher.cc

```diff
@@ -79,39 +53,86 @@ ss::future<> replicate_batcher::stop() {
-                i->_promise.set_exception(ss::gate_closed_exception());
+                i->set_exception(
+                  std::make_exception_ptr(ss::gate_closed_exception()));
             }
             _item_cache.clear();
         });
     });
 }
 
+ss::future<result<replicate_result>>
+replicate_batcher::cache_and_wait_for_result(
+  ss::promise<> enqueued,
+  std::optional<model::term_id> expected_term,
+  model::record_batch_reader r,
+  consistency_level consistency_lvl,
+  std::optional<std::chrono::milliseconds> timeout) {
+    try {
+        auto item = co_await do_cache(
+          expected_term, std::move(r), consistency_lvl, timeout);
+        auto ret_f = item->get_future();
+        // now request is already enqueued, we can release first
+        // stage future
+        enqueued.set_value();
+        try {
+            /**
+             * Dispatching flush in a background.
+             *
+             * The batcher mutex may be grabbed without the timeout as each item
+             * has its own timeout timer. The shutdown related exceptions may be
+             * ignored here as all pending item promises will be completed in
+             * replicate batcher stop method
+             *
+             */
+            ssx::background
+              = ssx::spawn_with_gate_then(_bg, [this]() -> ss::future<> {
+                    auto units = co_await _lock.get_units();
+                    co_await flush(std::move(units), false);
+                }).handle_exception([item](const std::exception_ptr& e) {
+                    item->set_exception(e);
+                });
+
+            co_return co_await std::move(ret_f);
+        } catch (...) {
+            co_return errc::replicate_batcher_cache_error;
+        }
+
+    } catch (...) {
+        enqueued.set_to_current_exception();
+        co_return errc::replicate_batcher_cache_error;
+    }
+}
+
 ss::future<replicate_batcher::item_ptr> replicate_batcher::do_cache(
   std::optional<model::term_id> expected_term,
   model::record_batch_reader&& r,
-  consistency_level consistency_lvl) {
-    return model::consume_reader_to_memory(std::move(r), model::no_timeout)
-      .then([this, expected_term, consistency_lvl](
-              ss::circular_buffer<model::record_batch> batches) {
-          ss::circular_buffer<model::record_batch> data;
-          size_t bytes = std::accumulate(
-            batches.cbegin(),
-            batches.cend(),
-            size_t{0},
-            [](size_t sum, const model::record_batch& b) {
-                return sum + b.size_bytes();
-            });
-          return do_cache_with_backpressure(
-            expected_term, std::move(batches), bytes, consistency_lvl);
+  consistency_level consistency_lvl,
+  std::optional<std::chrono::milliseconds> timeout) {
+    gate_guard guard(_bg);
+    auto batches = co_await model::consume_reader_to_memory(
+      std::move(r),
+      timeout ? model::timeout_clock::now() + *timeout : model::no_timeout);
```

Comment:
Can the timeout passed as a parameter to all of these methods be a non-optional `model::timeout::clock::time_point` so that you can just pass `model::no_timeout` instead of nullopt to indicate no timeout desired?

Reply:
there is a problem with this approach as the semaphore uses `steady_clock`. Conversion between the timepoints is impossible